### PR TITLE
Improvements to alt aliases

### PR DESCRIPTION
**Pull request by aaronraimist**
Attempt 2 at https://github.com/matrix-org/matrix-react-sdk/pull/4574

Step towards fixing https://github.com/vector-im/element-web/issues/13077

----

TR: Video

https://user-images.githubusercontent.com/5855073/122130924-ff628a00-cdfd-11eb-9100-c308f4067156.mov

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Improvements to alt aliases ([\#6180](https://github.com/matrix-org/matrix-react-sdk/pull/6180)). Contributed by @aaronraimist.

### CODE REVIEW

**Comment by t3chguy** [comment]
Could we get screenshots/gif of the changes?

**Comment by aaronraimist** [comment]
When I get back to the computer I’ll take some but it looks exactly like the original PR. It’s best experienced by using the Netlify preview or building it yourself.

**Comment by aaronraimist** [comment]
https://user-images.githubusercontent.com/5855073/122130924-ff628a00-cdfd-11eb-9100-c308f4067156.mov

File: src/components/views/room_settings/PublishedAliases.tsx [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+interface AEndCommit {
+    type: 'end_commit'
+}
+
+interface AAbortCommit {
+    type: 'abort_commit',
+    error: Error
+}
+
+type Action = AUpdateAlias | ASyncedAliases | ABeginCommit | AEndCommit | AAbortCommit;
```
Comment (turt2live) [inline src/components/views/room_settings/PublishedAliases.tsx:79]: I'm a bit worried that this is a lot of extra complexity... Can we get comments as to what these are trying to achieve, and what problems they solve?

Also, our interfaces are `INamed`, so in this case would be `IUpdateAliasAction` and such. The `type Action` should probably be named something that doesn't conflict with the dispatcher, such as `type AliasAction`

File: src/components/views/room_settings/PublishedAliases.tsx [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+/* This entire class exists so we can:
+ * 1. Add a tooltip to the field
+ * 2. Do validation
+ * TODO: Fixing up this, and AliasSettings, to not need a subclass is a very
+ * likely candidate for further cleanup
+ */
+
+class EditableAltAliasList extends EditableItemList {
+    private _publishedAliasField = React.createRef<Field>();
```
Comment (turt2live) [inline src/components/views/room_settings/PublishedAliases.tsx:151]: drop the underscore

File: src/components/views/room_settings/PublishedAliases.tsx [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+        default:
+            return state;
+    }
+}
+
+/* This entire class exists so we can:
+ * 1. Add a tooltip to the field
+ * 2. Do validation
+ * TODO: Fixing up this, and AliasSettings, to not need a subclass is a very
+ * likely candidate for further cleanup
```
Comment (turt2live) [inline src/components/views/room_settings/PublishedAliases.tsx:147]: open TODOs are a bit questionable

File: src/components/views/room_settings/PublishedAliases.tsx [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+        try {
+            await client.sendStateEvent(room.roomId, "m.room.canonical_alias", content, "");
+            if (isMounted.current) dispatch({type: END_COMMIT});
+        } catch (err) {
+            console.error(`Error updating aliases in ${room.roomId}`, err);
+
+            /* Don't throw up a dialog if the user closed the parent */
+            if (!isMounted.current) return;
+
+            dispatch({type: ABORT_COMMIT, error: err});
```
Comment (turt2live) [inline src/components/views/room_settings/PublishedAliases.tsx:252]: where is this dispatch function coming from?

File: src/components/views/room_settings/PublishedAliases.tsx [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+ * 1. Add a tooltip to the field
+ * 2. Do validation
+ * TODO: Fixing up this, and AliasSettings, to not need a subclass is a very
+ * likely candidate for further cleanup
+ */
+
+class EditableAltAliasList extends EditableItemList {
+    private _publishedAliasField = React.createRef<Field>();
+
+    _onAliasAdded = () => {
```
Comment (turt2live) [inline src/components/views/room_settings/PublishedAliases.tsx:153]: `private`, no underscore

File: src/components/views/room_settings/PublishedAliases.tsx [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+        this.props.onItemAdded(this.props.newItem);
+
+        this._publishedAliasField.current.focus();
+    };
+
+    _renderNewItemField() {
```
Comment (turt2live) [inline src/components/views/room_settings/PublishedAliases.tsx:159]: `private`, no underscore

**Review by turt2live — COMMENTED** [review]
Thank you for working on bringing this back. I realize that the stuff I'm commenting on is not necessarily your code, but I do still have concerns with the approach given it introduces a moderately complex state machine for a component which doesn't appear to need it.

I haven't personally tested this, but from the looks of it the majority of the usability concerns were worked out already. 

Design will probably want to take a look at this as well, as a sort of stopgap approach (hopefully). The spaces alias changes will almost certainly have collided with this too.

If it comes down to it, what is your interest in maintaining this and potentially pushing it forward with design support? Including potentially (re)writing a bunch of the state machine stuff.

**Comment by aaronraimist** [comment]
> If it comes down to it, what is your interest in maintaining this and potentially pushing it forward with design support? Including potentially (re)writing a bunch of the state machine stuff.

I am not super interested in rewriting it. I was mostly reintroducing this PR because it already passed a code review and was just missing a design review. If it needs to be rewritten then it'll probably have to wait for someone else to pick it up.

**Comment by Palid** [comment]
@aaronraimist I'm taking this one over, hopefully you don't mind. 😄

**Comment by turt2live** [comment]
@aaronraimist I think this got replaced with https://github.com/matrix-org/matrix-react-sdk/pull/7107 ? Are there remaining bits of this or the linked issue which might still need doing?

**Comment by aaronraimist** [comment]
@turt2live #7107 is an improvement but in my opinion it doesn't replace this PR. #7107 checks that the address about to be published looked like an alias but this PR actually checks if the alias points at the room and provides a helpful error if not.

Currently if you enter a valid looking alias but it doesn't point at the room, you get this error which doesn't explain much.


I'm not sure if @Palid wants to leave this open and build on top of it or close it and rewrite it in a different way.

**Comment by turt2live** [comment]
I think if you're able to pick it up instead that would be ideal.

**Comment by aaronraimist** [comment]
I'll see what I can do but can't promise anything in the near future

**Pull request closed without merging**